### PR TITLE
Feature/merging

### LINKED ISSUE #2450: Open mSupply Record Merging
## Is your feature request related to a problem? Please describe 👀

[Based on the Legacy mSupply KDD regarding merge strategy.](https://github.com/msupply-foundation/msupply/blob/6342dba1ac0aacf010078ce4478c58d67f545916/decisions/0005-sync-v5-merges.md)

Open mSupply must support record merging in the following areas:
 
- [x] #2451
- [x] #2550
- [x] #2551
- [x] #2552
- [x] #2553
- [x] #2639
- [x] #2554
- [x] #2620
- [x] #2626
- [x] #2635
- [x] #2636
- [x] #2637
- [x] #2452
- [x] #2555 
- [x] #2556
- [x] #2557 
- [x] #2558 
- [x] #2559
- [x] #2681
- [x] #2732 
- [x] #2735 
- [x] #2682
- [x] #2683
- [x] #2684 
- [x] #2754
- [x] #2561 
- [x] #2562  
- [x] #2606 
- [x] #2815
- [x] #2745 
- [x] #2746
- [x] #2747 
- [x] #2748  
- [x] #2749
- [x] #2750 
- [x] #2767
- [x] #2771
- [x] #2454
- [x] #2565 
- [x] #2566 
- [x] #2769
- [x] #2607 
- [x] #2698
- [x] #2733 
- [x] #2759
- [x] #2833
- [x] #2835 
- [x] #2836 
- [x] #2837 
- [x] #2838 
- [x] #2839 
- [x] #2840 
- [x] #2841 
- [x] #2842  
- [x] #2843 
- [x] #2844 
- [x] #2845 
- [x] #2846 
- [x] #2847 
- [x] #2848 
- [x] #2849 
- [x] #2850 
- [x] #2885

### Additional context 💌

Will be using `someTable_link` format for each merge-able table. An alternative would be `link` with `link.type = "someTable"` but it is more difficult to setup SQL FK constraints when done that way.

Additional areas that could require merging support in the future:

- #2453
- Store merging
- User merging 

### Moneyworks Jobcode 🧰

OMS:MERGE

### PR DESCRIPTION
**Pull request by Chris-Petty**
Fixes #2450

# 👩🏻‍💻 What does this PR do? 

Brings in the changes of item and name merging using `item_link` and `name_link` respectively. 

# 🧪 How has/should this change been tested? 

Good amount of automated test suites but for the manual pass:

- [ ] mSupply Desktop Central Server (MDCS) and an OMS android or desktop. Synced, but with OMS on an older version (without these changes - we want to ensure migration works correctly)
- [ ] Merge 2 items and merge 2 names on MDCS and sync. (This will ensure there are unprocessed merge records in the integration buffer of OMS)
- [ ] Migrate OMS to the latest version including these changes
- [ ] Check:
  - [ ] Item catalogue updated
  - [ ] Master lists updated
  - [ ] Stock take lines updated 
  - [ ] Stock lines updated
  - [ ] Requisition lines updated
  - [ ] In/outbound shipment lines and prescription lines NOT updated (the original name when the line was added should be preserved)
- [ ] Check the names appear merged in
  - [ ] Customers (My product knowledge betrays me, is there such a list in OMS UI?)
  - [ ] Suppliers
  - [ ] Shipments referencing merged name
  - [ ] Internal orders/Requisitions **Tricky, you probably don't have any to the merged name as you can't merge store names! And you can't make these records to regular names/external parties in OMS... yet**
  - [ ] Prescriptions
  - [ ] Barcodes
  - [ ] Patient encounters
- [ ] Do all of the above again merging the kept name into another name and the kept item into another item - checking that merge on regular sync works correctly, not just after migration.
- [ ] Spicy edge case: Make a customer invoice to a regular name. Merge that name into a store that is active on your OMS site. Sync. Login to that store and check that a supplier invoice has been created. (Note, you can in effect make a SI to your own store in doing this. Not ideal, but currently expected behaviour. Should be rare...)

## 💌 Any notes for the reviewer?

All the code here has been reviewed as it was merged into the feature branch, so detail isn't so important. We are still waiting on the integration tests to be fixed, problems identified should be done very soon.

## 📃 Documentation

Hmm nothing yet? Merging is still done by the user in legacy mSupply, so no real new functionality here.

### CODE REVIEW

**Review by roxy-dao — COMMENTED** [review]
gif

Just a few changes below before this can be merged to develop (: mostly to do with indexes that were removed in 1.6~

**Review by roxy-dao — CHANGES_REQUESTED** [review]
Forgot to change 😆

File: server/service/src/processors/transfer/shipment/mod.rs [hunk 1]
```diff
@@ -144,6 +144,8 @@ pub(crate) fn process_shipment_transfers(
                     .ok_or_else(|| Error::NameIsNotAnActiveStore(log.clone()))?,
             };
 
+            // MERGE: Ignore if invoice name_link_id points to store's name. Supplying to itself! (Can happen with names are merge into stores)
```
[the remark this answers is not in the document]
Comment (Chris-Petty) [reply server/service/src/processors/transfer/shipment/mod.rs:147]: Do we even do TODOs? 😅

File: server/repository/src/migrations/v1_06_00/indexes.rs [hunk 2]
```diff
@@ -15,7 +15,6 @@ pub(crate) fn migrate(connection: &StorageConnection) -> anyhow::Result<()> {
         CREATE INDEX index_contact_trace_patient_id ON contact_trace (patient_id);
         CREATE INDEX index_contact_trace_contact_patient_id ON contact_trace (contact_patient_id);
         CREATE INDEX index_contact_trace_store_id ON contact_trace (store_id);
-        CREATE INDEX index_stock_line_supplier_id ON stock_line (supplier_id);
```
[the remark this answers is not in the document]
Comment (Chris-Petty) [reply server/repository/src/migrations/v1_06_00/indexes.rs:18]: Nice catch!

File: server/service/src/processors/transfer/shipment/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (roxy-dao) [reply server/service/src/processors/transfer/shipment/mod.rs:147]: Eventually... maybe.. idk...

File: server/repository/src/db_diesel/clinician_link_row.rs [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+allow_tables_to_appear_in_same_query!(clinician_link, name_link);
+
+#[derive(Clone, Insertable, Queryable, Debug, PartialEq, AsChangeset, Eq)]
+#[table_name = "clinician_link"]
+pub struct ClinicianLinkRow {
+    pub id: String,
+    pub clinician_id: String,
+}
+
+impl Default for ClinicianLinkRow {
```
[the remark this answers is not in the document]
Comment (Chris-Petty) [reply server/repository/src/db_diesel/clinician_link_row.rs]: https://github.com/msupply-foundation/open-msupply/pull/2906/commits/fc5f65d88dfd74a65d36aa4ea4e102107d994691

File: server/repository/src/db_diesel/item_link_row.rs [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+allow_tables_to_appear_in_same_query!(item_link, name_link);
+
+#[derive(Clone, Insertable, Queryable, Debug, PartialEq, AsChangeset, Eq)]
+#[table_name = "item_link"]
+pub struct ItemLinkRow {
+    pub id: String,
+    pub item_id: String,
+}
+
+impl Default for ItemLinkRow {
```
[the remark this answers is not in the document]
Comment (Chris-Petty) [reply server/repository/src/db_diesel/item_link_row.rs]: https://github.com/msupply-foundation/open-msupply/pull/2906/commits/fc5f65d88dfd74a65d36aa4ea4e102107d994691

File: server/repository/src/migrations/v1_06_00/patient_id_indices.rs [hunk 5]
```diff
@@ -5,7 +5,6 @@ pub(crate) fn migrate(connection: &StorageConnection) -> anyhow::Result<()> {
         connection,
         r#"
         CREATE INDEX index_encounter_patient_id ON encounter (patient_id);
-        CREATE INDEX index_program_enrolment_patient_id ON program_enrolment (patient_id);
```
[the remark this answers is not in the document]
Comment (Chris-Petty) [reply server/repository/src/migrations/v1_06_00/patient_id_indices.rs:8]: https://github.com/msupply-foundation/open-msupply/pull/2906/commits/e6bfb3a729ee75da30b996f2af3657b7c40dc200

File: server/repository/src/migrations/v1_06_00/indexes.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Chris-Petty) [reply server/repository/src/migrations/v1_06_00/indexes.rs:18]: https://github.com/msupply-foundation/open-msupply/pull/2906/commits/e6bfb3a729ee75da30b996f2af3657b7c40dc200

File: server/repository/src/merging.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+Internally, the repository looks up all `name_links` which match the `name_id` and then uses these `name_links` to do the query.
+Note, if a merge happened while the service layer still uses the id of the soft deleted entity, a query using this id will return nothing because there will be no links pointing to this old id.
+
+<!-- However, updating a name_id is done using the row.name_link_id -->
+
+When updating an entity of a data row the link fields of the repository objects are used, e.g. `row.name_link_id = "New name id"`.
+This works because current entity has a matching link entry with the same id, e.g `name_link.id = name_id`.
+For example, when creating an outbound shipment you should use `invoice.name_link_id = customer_name.id` because `customer_name.id` is guaranteed to have a corresponding `name_link_id` row.
+For existing shipments you should **not** update `invoice.name_link_id` to the returned related name entity's `name.id`, even if it may be different.
+Any difference is due to a merge has occurred, and updating the `name_link_id` would undermine a design goal of potentially being able to "unmerge" entities.
```
[the remark this answers is not in the document]
Comment (Chris-Petty) [reply server/repository/src/merging.md]: Is this better?

https://github.com/msupply-foundation/open-msupply/pull/2906/commits/b35e6a4499e30435ba5d92f22913f070f84f9ae1

[the remark this answers is not in the document]
Comment (roxy-dao) [reply server/repository/src/merging.md]: Heaps! Thanks ❤️

**Review by roxy-dao — APPROVED** [review]
Woooot~ thanks for the changes Chris (:

**Comment by Chris-Petty** [comment]
One of the scarier merge button presses...

**Pull request merged**